net/netcheck/netcheck.go

```diff
@@ -1119,23 +1123,37 @@ func (c *Client) addReportHistoryAndSetPreferredDERP(r *Report) {
 			delete(c.prev, t)
 			continue
 		}
-		for hp, d := range pr.RegionLatency {
-			if bd, ok := bestRecent[hp]; !ok || d < bd {
-				bestRecent[hp] = d
+		for regionID, d := range pr.RegionLatency {
+			if bd, ok := bestRecent[regionID]; !ok || d < bd {
+				bestRecent[regionID] = d
 			}
 		}
 	}
 
 	// Then, pick which currently-alive DERP server from the
 	// current report has the best latency over the past maxAge.
 	var bestAny time.Duration
-	for hp := range r.RegionLatency {
-		best := bestRecent[hp]
+	var oldRegionCurLatency time.Duration
+	for regionID, d := range r.RegionLatency {
+		if regionID == prevDERP {
+			oldRegionCurLatency = d
+		}
+		best := bestRecent[regionID]
 		if r.PreferredDERP == 0 || best < bestAny {
 			bestAny = best
-			r.PreferredDERP = hp
+			r.PreferredDERP = regionID
 		}
 	}
+
+	// If we're changing our preferred DERP but the old one's still
+	// accessible and the new one's not much better, just stick with
+	// where we are.
+	if prevDERP != 0 &&
+		r.PreferredDERP != prevDERP &&
+		oldRegionCurLatency != 0 &&
+		bestAny > oldRegionCurLatency/3*2 {
```

Comment:
It's a bit weird to compare latencies across time. A local latency change (e.g. moving farther from the wifi router) could make "time to SFO now" vs "time to NYC before" not really the right comparison.

Maybe that's fine in practice. Or maybe we should be comparing "advantage of SFO over NYC now" against "preference for stable DERP".

Reply:
Another approach here is to use a decay factor along with previous measurements to calculate DERP latency. That should make the measurements more stable, which by itself should reduce flip-flopping. And then we could use a much smaller hysteresis threshold (5%? 10?). 50% is really pretty significant.

Reply:
Well, it's already comparing the best latency sample in the past 5 minutes per region. That was the old logic. So it's not like it's just comparing the most recent ping latency data point.

Feel free to own this area if you have time to think about it & write tests, though. I'm not attached to it.

I just wanted to fix all the flip flopping we've been in some recent bug reports.